core/state/access_witness.go

```diff
@@ -89,83 +91,111 @@ func (aw *AccessWitness) Copy() *AccessWitness {
 // a contract creation.
-func (aw *AccessWitness) TouchAndChargeContractCreateInit(addr []byte, availableGas uint64) (uint64, uint64) {
-	gas1, wanted1 := aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BasicDataLeafKey, true, availableGas)
-	gas2, wanted2 := aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.CodeHashLeafKey, true, availableGas-gas1)
-	return gas1 + gas2, wanted1 + wanted2
+func (aw *AccessWitness) TouchContractCreateInit(addr []byte, availableGas uint64) {
+	aw.touchLocation(addr, zeroTreeIndex, utils.BasicDataLeafKey, true)
+	aw.touchLocation(addr, zeroTreeIndex, utils.CodeHashLeafKey, true)
 }
 
 func (aw *AccessWitness) TouchTxOriginAndComputeGas(originAddr []byte) {
 	for i := utils.BasicDataLeafKey; i <= utils.CodeHashLeafKey; i++ {
-		aw.touchAddressAndChargeGas(originAddr, zeroTreeIndex, byte(i), i == utils.BasicDataLeafKey, math.MaxUint64)
+		aw.touchLocation(originAddr, zeroTreeIndex, byte(i), i == utils.BasicDataLeafKey)
 	}
 }
 
 func (aw *AccessWitness) TouchTxTarget(targetAddr []byte, sendsValue, doesntExist bool) {
-	aw.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, utils.BasicDataLeafKey, sendsValue, math.MaxUint64)
+	aw.touchLocation(targetAddr, zeroTreeIndex, utils.BasicDataLeafKey, sendsValue)
 	// Note that we do a write-event in CodeHash without distinguishing if the tx target account
 	// exists or not. Pre-7702, there's no situation in which an existing codeHash can be mutated, thus
 	// doing a write-event shouldn't cause an observable difference in gas usage.
 	// TODO(7702): re-check this in the spec and implementation to be sure is a correct solution after
 	// EIP-7702 is implemented.
-	aw.touchAddressAndChargeGas(targetAddr, zeroTreeIndex, utils.CodeHashLeafKey, doesntExist, math.MaxUint64)
+	aw.touchLocation(targetAddr, zeroTreeIndex, utils.CodeHashLeafKey, doesntExist)
 }
 
-func (aw *AccessWitness) TouchSlotAndChargeGas(addr []byte, slot common.Hash, isWrite bool, availableGas uint64, warmCostCharging bool) uint64 {
+func (aw *AccessWitness) SlotGas(addr []byte, slot common.Hash, isWrite bool, warmCostCharging bool) uint64 {
 	treeIndex, subIndex := utils.GetTreeKeyStorageSlotTreeIndexes(slot.Bytes())
-	_, wanted := aw.touchAddressAndChargeGas(addr, *treeIndex, subIndex, isWrite, availableGas)
+	wanted := aw.calculateWitnessGas(addr, *treeIndex, subIndex, isWrite)
 	if wanted == 0 && warmCostCharging {
 		wanted = params.WarmStorageReadCostEIP2929
 	}
 	return wanted
 }
 
-func (aw *AccessWitness) touchAddressAndChargeGas(addr []byte, treeIndex uint256.Int, subIndex byte, isWrite bool, availableGas uint64) (uint64, uint64) {
+func (aw *AccessWitness) TouchSlot(addr []byte, slot common.Hash, isWrite bool) {
+	treeIndex, subIndex := utils.GetTreeKeyStorageSlotTreeIndexes(slot.Bytes())
+	aw.touchLocation(addr, *treeIndex, subIndex, isWrite)
+}
+
+func (aw *AccessWitness) calculateWitnessGas(addr []byte, treeIndex uint256.Int, subIndex byte, isWrite bool) uint64 {
```

Comment:
there's an issue with this approach, because if we have two grouped access, any access won't be recorded (i.e. a branch opening won't be recorded as happening, since this is let to `Touch`). As a result, the branch opening costs is charged again.

This means that e.g. for `ContractCreateCheckGas`, which accesses 2 leaves of the same group, it doesn't return 1900 + 200 + 200 = 2300, but 2x(1900 + 200) = 4200. So we need to change the interface to `calculateGas` so that the branch is computed once and for all. I have to check if variadic accesses are expensive in go.

Reply:
ugh, it seems that indeed, this will turn into a slice allocation (although it can be optimized by the compiler and passed onto the stack for a small number of arguments). I guess this should be tested, and also the performance needs to be measured.